processing/src/main/java/io/druid/query/QueryPlus.java

```diff
@@ -0,0 +1,104 @@
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query;
+
+import com.google.common.base.Preconditions;
+import io.druid.java.util.common.guava.Sequence;
+import io.druid.query.spec.QuerySegmentSpec;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+
+/**
+ * An immutable composite object of {@link Query} + extra stuff needed in {@link QueryRunner}s. This "extra stuff"
+ * is only {@link QueryMetrics} yet.
+ */
+public final class QueryPlus<T>
+{
+  /**
+   * Returns the minimum bare QueryPlus object with the given query. {@link #getQueryMetrics()} of the QueryPlus object,
+   * returned from this factory method, returns {@code null}.
+   */
+  public static <T> QueryPlus<T> wrap(Query<T> query)
+  {
+    Preconditions.checkNotNull(query);
+    return new QueryPlus<>(query, null);
+  }
+
+  private final Query<T> query;
+  private final QueryMetrics<?> queryMetrics;
+
+  private QueryPlus(Query<T> query, QueryMetrics<?> queryMetrics)
+  {
+    this.query = query;
+    this.queryMetrics = queryMetrics;
+  }
+
+  public Query<T> getQuery()
+  {
+    return query;
+  }
+
+  @Nullable
+  public QueryMetrics<?> getQueryMetrics()
+  {
+    return queryMetrics;
+  }
+
+  /**
+   * Returns the same QueryPlus object, if it already has {@link QueryMetrics} ({@link #getQueryMetrics()} returns not
+   * null), or returns a new QueryPlus object with {@link Query} from this QueryPlus and QueryMetrics created using the
+   * given {@link QueryToolChest}, via {@link QueryToolChest#makeMetrics(Query)} method.
+   */
+  public QueryPlus<T> withQueryMetrics(QueryToolChest<T, ? extends Query<T>> queryToolChest)
+  {
+    if (queryMetrics != null) {
+      return this;
+    } else {
+      return new QueryPlus<>(query, ((QueryToolChest) queryToolChest).makeMetrics(query));
+    }
+  }
+
+  /**
+   * Equivalent of withQuery(getQuery().withQuerySegmentSpec(spec)).
+   */
+  public QueryPlus<T> withQuerySegmentSpec(QuerySegmentSpec spec)
+  {
+    return new QueryPlus<>(query.withQuerySegmentSpec(spec), queryMetrics);
+  }
+
+  /**
+   * Returns a QueryPlus object with {@link QueryMetrics} from this QueryPlus object, and the provided {@link Query}.
+   */
+  public QueryPlus<T> withQuery(Query<T> replacementQuery)
+  {
+    return new QueryPlus<>(replacementQuery, queryMetrics);
+  }
+
+  public Sequence<T> run(QuerySegmentWalker walker, Map<String, Object> context)
+  {
+    if (query instanceof BaseQuery) {
+      return ((BaseQuery) query).getQuerySegmentSpec().lookup(query, walker).run(this, context);
```

Comment:
It means that `Query` and `BaseQuery` should learn about `QueryPlus`, that I didn't want to happen.